src/DynamoCoreWpf/ViewModels/Core/InPortViewModel.cs

```diff
@@ -0,0 +1,306 @@
+            {
+                _port.UsingDefaultValue = value;
+            }
+        }
+
+        /// <summary>
+        /// If should display Use Levels popup menu. 
+        /// </summary>
+        public bool ShowUseLevelMenu
+        {
+            get
+            {
+                return _showUseLevelMenu;
+            }
+            set
+            {
+                _showUseLevelMenu = value;
+                RaisePropertyChanged("ShowUseLevelMenu");
+            }
+        }
+
+        /// <summary>
+        /// If UseLevel is enabled on this port.
+        /// </summary>
+        public bool UseLevels
+        {
+            get { return _port.UseLevels; }
+        }
+
+        /// <summary>
+        /// Determines whether or not the UseLevelsSpinner is visible on the port.
+        /// </summary>
+        public Visibility UseLevelSpinnerVisible
+        {
+            get
+            {
+                if (PortType == PortType.Output) return Visibility.Collapsed;
+                if (UseLevels) return Visibility.Visible;
+                return Visibility.Hidden;
+            }
+        }
+
+        /// <summary>
+        /// If should keep list structure on this port.
+        /// </summary>
+        public bool ShouldKeepListStructure
+        {
+            get { return _port.KeepListStructure; }
+        }
+
+        /// <summary>
+        /// Levle of list.
+        /// </summary>
+        public int Level
+        {
+            get { return _port.Level; }
+            set
+            {
+                ChangeLevel(value);
+            }
+        }
+
+        /// <summary>
+        /// The visibility of Use Levels menu.
+        /// </summary>
+        public Visibility UseLevelVisibility
+        {
+            get
+            {
+                if (_node.ArgumentLacing != LacingStrategy.Disabled)
+                {
+                    return Visibility.Visible;
+                }
+                else
+                {
+                    return Visibility.Collapsed;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Sets the color of the small rectangular marker on each input port.
+        /// </summary>
+        public SolidColorBrush PortValueMarkerColor
+        {
+            get => portValueMarkerColor;
+            set
+            {
+                portValueMarkerColor = value;
+                RaisePropertyChanged(nameof(PortValueMarkerColor));
+            }
+        }
+
+        /// <summary>
+        /// Sets the color of the use levels popup in the input port context menu.
+        /// This changes when the Keep List Structure option is activated and the port
+        /// is connected, upon which it turns blue.
+        /// </summary>
+        public SolidColorBrush UseLevelsMenuColor
+        {
+            get
+            {
+                return ShouldKeepListStructure && _port.IsConnected
+                    ? new SolidColorBrush(Color.FromArgb(255, 60, 60, 60))
```

Comment:
@saintentropy cache or recycle these colors please instead of creating new ones.